### PR TITLE
feat: support for !includesub directive

### LINKED ISSUE #501: Feature request: Support for !includesub
I guess this could build on the already existing support for `!include`, adding a few lines of code to extract the appropriate part of the included file.

### PR DESCRIPTION
**Pull request by kumachan-mis**
close #501

### CODE REVIEW

**Comment by kumachan-mis** [comment]
The following issues are remaining, but they will be out of this PR
- include(sub) loop detection
- caching for multiple-times including

File: src/finder/finderUtil.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+export function extractSubIncludedText(content: string, tag: string): string {
+  const STARTSUB_REGEX = /^\s*!startsub\s+(.*)\s*$/;
```
Comment (exKAZUu) [inline src/finder/finderUtil.ts:2]: Should move constant defintions out of the function due to performance.

File: src/finder/finderUtil.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  const ENDSUB_REGEX = /^\s*!endsub\s*$/;
+
+  const contentLines = content.split('\n');
+  const subTextLines = [];
+  let subDepth = 0;
+  for (const line of contentLines) {
+    if (subDepth === 0) {
+      const match = STARTSUB_REGEX.exec(line);
+      if (match !== null && match[1] === tag) subDepth++;
```
Comment (exKAZUu) [inline src/finder/finderUtil.ts:11]: I think this should be `if (match && match[1] === tag)` (and similar conditinos should be simplified).

File: src/finder/gitHubFinder.ts [hunk 3]
```diff
@@ -28,44 +29,81 @@ export class GitHubCodeBlockFinder implements CodeFinder {
...
+    const dirUrl = webPageUrl.replace(/\/[^/]*\.(plantuml|pu|puml|wsd)(\?.*)?$/, '');
+
+    const preprocessedLines = [];
+    for (const line of fileTextLines) {
+      const match = this.INCLUDE_REGEX.exec(line);
+      if (!match) {
+        preprocessedLines.push(line);
+        continue;
+      }
+      const includedText = await (async () => {
```
Comment (exKAZUu) [inline src/finder/gitHubFinder.ts:68]: I feel this block is a bit complicated (though I can understand `return` is useful).
How about the following code.

```
        const includedFileUrl = `${dirUrl}/${match[1]}`;
        const response = await fetch(includedFileUrl);
        let includedText = '';
        if (response.ok) {
          const htmlString = await response.text();
          const $body = $(new DOMParser().parseFromString(htmlString, 'text/html')).find('body');
          const fileTexts = await this.find(includedFileUrl, $body);
          includedText = fileTexts.map((fileText) => fileText.text.replace(/@startuml/g, '').replace(/@enduml/g, '')).join('\n');
        }
```

Comment (exKAZUu) [reply src/finder/gitHubFinder.ts:68]: Or can we rewrite this block as follows?

```
        const includedFileUrl = `${dirUrl}/${match[1]}`;
        const response = await fetch(includedFileUrl);
        if (response.ok) {
          const htmlString = await response.text();
          const $body = $(new DOMParser().parseFromString(htmlString, 'text/html')).find('body');
          const fileTexts = await this.find(includedFileUrl, $body);
          const includedText = fileTexts.map((fileText) => fileText.text.replace(/@startuml/g, '').replace(/@enduml/g, '')).join('\n');
          preprocessedLines.push(includedText);
        }
```

File: src/finder/gitHubFinder.ts [hunk 4]
```diff
@@ -28,44 +29,83 @@ export class GitHubCodeBlockFinder implements CodeFinder {
...
+    for (const line of fileTextLines) {
+      const match = this.INCLUDE_REGEX.exec(line);
+      if (!match) {
+        preprocessedLines.push(line);
+        continue;
+      }
+
+      const includedFileUrl = `${dirUrl}/${match[1]}`;
+      const response = await fetch(includedFileUrl);
+      if (response.ok) {
```
Comment (exKAZUu) [inline src/finder/gitHubFinder.ts:71]: I think this code is cleaner than previous one :+1:

Could you rewrite `if (response.ok) { ... }` to `if (!response.ok) continue;`?
In the current code, we can reduce nesting depth.

**Review by exKAZUu — APPROVED** [review]
LGTM :+1:

**Pull request merged**